irohad/ametsuchi/impl/postgres_command_executor.cpp

```diff
@@ -528,228 +318,593 @@ namespace iroha {
+                                WHERE value < 2::decimal ^ (256 - $5)
+                                LIMIT 1) %s
+                  )
+                  ON CONFLICT (account_id, asset_id)
+                  DO UPDATE SET amount = EXCLUDED.amount
+                  RETURNING (1)
+               )
+          SELECT CASE
+              WHEN EXISTS (SELECT * FROM insert_dest LIMIT 1) THEN 0
+              %s
+              WHEN NOT EXISTS (SELECT * FROM has_dest_account LIMIT 1) THEN 2
+              WHEN NOT EXISTS (SELECT * FROM has_src_account LIMIT 1) THEN 3
+              WHEN NOT EXISTS (SELECT * FROM has_asset LIMIT 1) THEN 4
+              WHEN NOT EXISTS (SELECT value FROM new_src_value
+                               WHERE value >= 0 LIMIT 1) THEN 5
+              WHEN NOT EXISTS (SELECT value FROM new_dest_value
+                               WHERE value < 2::decimal ^ (256 - $5)
+                               LIMIT 1) THEN 6
+              ELSE 7
+          END AS result)";
+
+    std::string CommandError::toString() const {
+      return (boost::format("%s: %s") % command_name % error_message).str();
+    }
+
+    PostgresCommandExecutor::PostgresCommandExecutor(soci::session &sql)
+        : sql_(sql), do_validation_(true) {}
+
+    void PostgresCommandExecutor::setCreatorAccountId(
+        const shared_model::interface::types::AccountIdType
+            &creator_account_id) {
+      creator_account_id_ = creator_account_id;
+    }
+
+    void PostgresCommandExecutor::doValidation(bool do_validation) {
+      do_validation_ = do_validation;
+    }
+
+    CommandResult PostgresCommandExecutor::operator()(
+        const shared_model::interface::AddAssetQuantity &command) {
+      auto &account_id = creator_account_id_;
+      auto &asset_id = command.assetId();
+      auto amount = command.amount().toStringRepr();
+      int precision = command.amount().precision();
+
+      auto cmd = boost::format("EXECUTE %1% ('%2%', '%3%', %4%, '%5%')");
+
+      if (do_validation_) {
+        cmd = (cmd % "addAssetQuantityWithValidation");
+      } else {
+        cmd = (cmd % "addAssetQuantityWithoutValidation");
+      }
+
+      cmd = (cmd % account_id % asset_id % precision % amount);
+
+      std::vector<std::function<std::string()>> message_gen = {
+          [&] {
+            return missRolePerm(
+                creator_account_id_,
+                shared_model::interface::permissions::Role::kAddAssetQty);
+          },
+          [] { return std::string("Account does not exist"); },
+          [] {
+            return std::string("Asset with given precision does not exist");
+          },
+          [] { return std::string("Summation overflows uint256"); },
+      };
+      return executeQuery(sql_, cmd.str(), "AddAssetQuantity", message_gen);
+    }
+
+    CommandResult PostgresCommandExecutor::operator()(
+        const shared_model::interface::AddPeer &command) {
+      auto &peer = command.peer();
```

Comment:
Each execute functions contains steps:
1. convert params to string
2. prepare execture string
3. add handlers
4. combine all previous.
Maybe we may create one function which will include all those steps and operator() will pass correct params to this function?

Reply:
Added TODO